functional-samples/tutorial.google-analytics/scripts/google-analytics.js

```diff
@@ -0,0 +1,122 @@
+    let clientId = result.clientId;
+    if (!clientId) {
+      // Generate a unique client ID, the actual value is not relevant
+      clientId = self.crypto.randomUUID();
+      await chrome.storage.local.set({ clientId });
+    }
+    return clientId;
+  }
+
+  // Returns the current session ID, or creates a new one if one doesn't exist or
+  // the previous one has expired.
+  async getOrCreateSessionId() {
+    // Store session in memory storage
+    const result = await chrome.storage.session.get('sessionData');
+    let sessionData = result.sessionData;
+    // Check if session exists and is still valid
+    const currentTimeInMs = Date.now();
+    if (sessionData && sessionData.timestamp) {
+      // Calculate how long ago the session was last updated
+      const durationInMin = (currentTimeInMs - sessionData.timestamp) / 60000;
+      // Check if last update lays past the session expiration threshold
+      if (durationInMin > SESSION_EXPIRATION_IN_MIN) {
+        // Delete old session id to start a new session
+        sessionData = null;
+      } else {
+        // Update timestamp to keep session alive
+        sessionData.timestamp = currentTimeInMs;
+        await chrome.storage.session.set({ sessionData });
+      }
+    }
+    if (!sessionData) {
+      // Create and store a new session
+      sessionData = {
+        session_id: currentTimeInMs.toString(),
+        timestamp: currentTimeInMs.toString()
+      };
+      await chrome.storage.session.set({ sessionData });
+    }
+    return sessionData.session_id;
+  }
+
+  // Fires an event with optional params. Event names must only include letters and underscores.
+  async fireEvent(name, params = {}) {
+    // Configure session id and engagement time if not present, for more details see:
+    // https://developers.google.com/analytics/devguides/collection/protocol/ga4/sending-events?client_type=gtag#recommended_parameters_for_reports
+    if (!params.session_id) {
+      params.session_id = await this.getOrCreateSessionId();
+    }
+    if (!params.engagement_time_msec) {
+      params.engagement_time_msec = DEFAULT_ENGAGEMENT_TIME_MSEC;
+    }
+
+    try {
+      const response = await fetch(
+        `${
+          this.debug ? GA_DEBUG_ENDPOINT : GA_ENDPOINT
+        }?measurement_id=${MEASUREMENT_ID}&api_secret=${API_SECRET}`,
+        {
+          method: 'POST',
+          body: JSON.stringify({
+            client_id: await this.getOrCreateClientId(),
+            events: [
+              {
+                name,
+                params
+              }
+            ]
+          })
+        }
+      );
+      if (!this.debug) {
+        return;
+      }
+      console.log(await response.text());
+    } catch (e) {
+      console.log('Google Analytics request failed with an exception', e);
+    }
+  }
+
+  // Fire a page view event.
+  async firePageViewEvent(title, location, additionalParams = {}) {
+    return this.fireEvent('page_view', {
+      page_title: title,
```

Comment:
nit: maybe use `page_title` and `page_location` as the variable name, for consistency sake from lines 85-86